src/build_system.ml

```diff
@@ -888,6 +887,99 @@ The following targets are not:
           end
         end)
 
+  (** If both [a] and [a/b] are source directories, we don't allow the rules for
+      [a] to define generated directories under [a/b] (e.g.
+      [a/b/.generated-by-a]).
+
+      The purpose is to avoid dependency cycles when computing the list of
+      subdirectories of [b]: you'd need to load rules for [a], for which you
+      often need to load rules for [a/b], at which point you need to do stale
+      artifact deletion, so you need to have computed the set of children
+      of [b] already.
+
+      One reasonable alternative is to delay stale artifact deletion until
+      it's actually necessary and I (aalekseyev) believe it to be a better
+      approach, but for now this restriction gives us an easier and more
+      incremental way forward.
+
+      This is in addition to another more general restriction:
+      a directory is only allowed to be generated if its parent knows
+      about it.
+
+      This module encodes those restrictions. *)
+  module Generated_directory_restrictions : sig
+
+    type restriction =
+      | Unrestricted
+      | Restricted of Path.Unspecified.w Dir_set.t Memo.Lazy.t
+
+    (** Used by the child to ask about the restrictions placed by the parent. *)
+    val allowed_by_parent
+      : dir:Path.Build.t -> restriction
+
+    (** Used by the parent to check what are the subdirs that it's forbidden
+        from generating rules in. *)
+    val is_forbidden_generated_subdir :
+      dir:Path.Build.t -> subdir:Path.Build.t -> bool
+
+  end = struct
+
+    type restriction =
+      | Unrestricted
+      | Restricted of Path.Unspecified.w Dir_set.t Memo.Lazy.t
+
+    let corresponding_source_dir ~dir =
+      let t = t () in
+      match Utils.analyse_target dir with
+      | Install _ | Alias _ | Other _ ->
+        None
+      | Regular (_ctx, sub_dir) ->
+        File_tree.find_dir t.file_tree sub_dir
+
+    let source_subdirs_of_build_dir ~dir =
+      match corresponding_source_dir ~dir with
+      | None -> String.Set.empty
+      | Some dir ->
+        File_tree.Dir.sub_dir_names dir
+
+    let is_forbidden_generated_subdir ~dir ~subdir =
+      match Path.Local_gen.descendant ~of_:dir subdir with
+      | None -> false
+      | Some reach ->
+        match Path.Local_gen.split_first_component reach with
+        | None ->
+          (* allowed to generate rules inside itself *)
+          false
+        | Some (child, _) ->
+          if Option.is_none (
+            corresponding_source_dir ~dir:(Path.Local_gen.relative dir child))
+          then
+            (* allowed to generate directories inside itself *)
+            false
+          else
+            (* allowed to generate rules in child directories as long as the
+               directory itself is not generated *)
+            Option.is_none (corresponding_source_dir ~dir:subdir)
```

Comment:
I'm having trouble following the logic in this last branch. Let's consider the following example:

- dir = a
- subdir = a/b/c
- a/b/c is a source dir

Then, if I read the code correctly we end up in this branch which returns `false`, meaning that `a` is allowed to generate directories in a/b/c. I thought this was disallowed. BTW, I'd personally use `is_allowed` rather than `is_forbidden` as I find it easier to reason about allowed things rather than disallowed things

Reply:
It's OK to generate rules in descendant source dirs directly. (or at least it's OK as far as the existence of the directory is concerned)
In fact there's a coq test that relies on this.

What's disallowed in this feature is generated *directories* under other source directories.

Good point on "allowed" vs "forbidden". I'll change that.

Reply:
ok